clients/application-amqp/src/main/java/org/eclipse/hono/application/client/amqp/ProtonBasedApplicationClient.java

```diff
@@ -247,55 +265,69 @@ public final void disconnect(final Handler<AsyncResult<Void>> completionHandler)
-                @Override
-                public Future<Void> close() {
-                    return recv.close();
-                }
-            });
+                }, 
+                closeHandler);
     }
 
     private Future<MessageConsumer> createAsyncConsumer(
             final String sourceAddress,
             final Function<DownstreamMessage<AmqpMessageContext>, Future<Void>> messageHandler,
             final Handler<Throwable> closeHandler) {
 
-        return GenericReceiverLink.create(
-                connection,
+        return createConsumer(
                 sourceAddress,
                 (delivery, message) -> {
                     try {
                         final var msg = ProtonBasedDownstreamMessage.from(message, delivery);
                         messageHandler.apply(msg)
-                            .onSuccess(ok -> {
-                                if (!delivery.isSettled()) {
-                                    LOG.debug("client provided message handler did not settle message, auto-accepting ...");
-                                    ProtonHelper.accepted(delivery, true);
-                                }
-                            })
-                            .onFailure(t -> {
-                                handleException(t, delivery);
-                            });
+                                .onSuccess(ok -> {
+                                    if (!delivery.isSettled()) {
+                                        LOG.debug("client provided message handler did not settle message, auto-accepting ...");
+                                        ProtonHelper.accepted(delivery, true);
+                                    }
+                                })
+                                .onFailure(t -> handleMessageHandlerError(t, delivery));
                     } catch (final Throwable t) {
-                        handleException(t, delivery);
+                        handleMessageHandlerError(t, delivery);
                     }
                 },
-                false,
-                s -> Optional.ofNullable(closeHandler).ifPresent(h -> h.handle(null)))
-            .map(recv -> new MessageConsumer() {
-                @Override
-                public Future<Void> close() {
-                    return recv.close();
-                }
-            });
+                closeHandler);
+    }
+
+    private Future<MessageConsumer> createConsumer(
+            final String sourceAddress,
+            final BiConsumer<ProtonDelivery, Message> messageConsumer,
+            final Handler<Throwable> closeHandler) {
+
+        // wrap the handler to ensure distinct objects are added to the consumerCloseHandlers list
+        // preventing issues if the same closeHandler instance is used for multiple createConsumer() invocations
+        final Handler<Throwable> wrappedCloseHandler = closeHandler != null ? (thr -> closeHandler.handle(thr)) : null;
```

Comment:
what difference does this make?

Reply:
> preventing issues if the same closeHandler instance is used for multiple createConsumer() invocations

The scenario in which there is an issue is a bit theoretical:
- The same `closeHandler` instance is used with 2 `createConsumer` invocations.
- One consumer instance is closed explicitly via the `MessageConsumer#close` method, and for some reason the `close` method is invoked twice for that consumer.
- Then there is no closeHandler anymore in the `consumerCloseHandlers` list to be invoked for the _other_ consumer instance.

In more general terms, the aim here is to have distinct `consumerCloseHandlers` list entries for each `createConsumer` invocation, so that `consumerCloseHandlers.remove(handler)` doesn't remove the wrong handler.